staking/staking-specification.md

```diff
@@ -0,0 +1,815 @@
+
+A pool with _D_ delegated stake that earned _R_ rewards for its pool members in a given epoch, the reward (_r_) for a member that delegated _d_ stake is computed by:
+
+<p align="center"><img src="./assets/reward_tracking/RewardSingleEpoch.png" height="60" /></p>
+
+
+The member's reward after n epochs (given member does not change their stake) is then given by:
+
+<p align="center"><img src="./assets/reward_tracking/RewardAfterManyEpochs.png" height="60" /></p>
+
+
+When a member modifies their stake in the pool, the [StoredBalance struct](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/interfaces/IStructs.sol) gives us:
+1. How many epochs they were staked (`n`)
+2. How much stake they had contributed during those epochs (`d`)
+
+In addition to these values, we also need sum of ratios `R_k / D_k`, for each epoch `k` that the member was delegated. This ratio is available during the pool's finalization of epoch `k`. We are able to do store this information concisely using a cumulative sum of these reward ratios, as follows:
+
+We store the following ratio for each epoch that a reward is earned for the pool:
+<p align="center"><img src="./assets/reward_tracking/WhatWeStore.png" height="60" /></p>
+
+We compute a member's reward using the following equation:
+<p align="center"><img src="./assets/reward_tracking/RewardFromWhatWeStore.png" height="40" /></p>
+
+Example:
+
+<p align="center"><img src="./assets/reward_tracking/RewardFromWhatWeStore-Example.png" height="60" /></p>
+
+This cumulative reward along with the stored balance of a member, we are able to compute their reward in the pool at any time.
+
+This information is stored on-chain as follows:
+```
+// mapping from Owner to Pool Id to Amount Delegated
+mapping (address  =>  mapping (bytes32  => IStructs.StoredBalance)) internal _delegatedStakeToPoolByOwner;
+
+// mapping from Pool Id to Amount Delegated
+mapping (bytes32  => IStructs.StoredBalance) internal _delegatedStakeByPoolId;
+
+// mapping from Pool Id to Epoch to Reward Ratio
+mapping (bytes32  =>  mapping (uint256  => IStructs.Fraction)) internal _cumulativeRewardsByPool;
+```
+
+#### 9.3.1 Computing Rewards - in Practice
+
+In the equations above, a staker earned rewards from epochs `[0..n]`. This means that the staker undelegated during epoch `n` and stopped earning rewards in epoch `n+1`. So at the time of the call, we don't have access to the reward for epoch `n`.
+
+In practice, this equation becomes:
+<p align="center"><img src="./assets/reward_tracking/RewardAfterManyEpochs-InPractice.png" height="60" /></p>
+
+
+The final equation for computing a member's reward during epoch `n` becomes:
+
+<p align="center"><img src="./assets/reward_tracking/Reward-Final.png" height="60" /></p>
+
+
+#### 9.3.2 Handling Epochs With No Rewards
```

Comment:
Thanks for writing this section. I think this provides good clarification on why having a sparse `cumulativeRewardsByPool` mapping is not an issue.